x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregatorFactory.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.analytics.topmetrics;
+
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories.Builder;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.MultiValuesSourceFieldConfig;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.sort.BucketedSort;
+import org.elasticsearch.search.sort.SortBuilder;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class TopMetricsAggregatorFactory extends AggregatorFactory {
+    private final List<SortBuilder<?>> sortBuilders;
+    private final MultiValuesSourceFieldConfig metricField;
+
+    public TopMetricsAggregatorFactory(String name, QueryShardContext queryShardContext, AggregatorFactory parent,
+            Builder subFactoriesBuilder, Map<String, Object> metaData, List<SortBuilder<?>> sortBuilders,
+            MultiValuesSourceFieldConfig metricField) throws IOException {
+        super(name, queryShardContext, parent, subFactoriesBuilder, metaData);
+        this.sortBuilders = sortBuilders;
+        this.metricField = metricField;
+    }
+
+    @Override
+    protected TopMetricsAggregator createInternal(SearchContext searchContext, Aggregator parent, boolean collectsFromSingleBucket,
+            List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+        ValuesSourceConfig<ValuesSource.Numeric> metricFieldSource = ValuesSourceConfig.resolve(queryShardContext, ValueType.NUMERIC,
+                metricField.getFieldName(), metricField.getScript(), metricField.getMissing(), metricField.getTimeZone(), null);
+        ValuesSource.Numeric metricValueSource = metricFieldSource.toValuesSource(queryShardContext);
```

Comment:
I understand why we're not using `MultiValuesSourceAggregationBuilder`, but I don't understand why we aren't using `ValuesSourceAggregationBuilder` (and `ValuesSourceAggregatorFactory`).  We have one field that we want to get a `ValuesSource` for, which is what `ValuesSourceAggregationBuilder` does.  Maybe there's some reason I'm missing here?  Using VSAB will also make my life easier when it comes time to wire this up to the registry ;)

Reply:
> We have one field that we want to get a `ValuesSource` for, which is what `ValuesSourceAggregationBuilder` does. Maybe there's some reason I'm missing here? Using VSAB will also make my life easier when it comes time to wire this up to the registry ;)

I'll look at it and let you know. I started with the multi-value source config thing but it had issues with the request format. Right now we have a single value here but I believe we'll end up with more in the future.

Reply:
I just checked - `ValuesSourceAggregationBuilder` ties me to a single value source but I expect to support multiple values.